### PR TITLE
test: add LastCheckpointHintState (Present/Missing/Stale) for TestTableBuilder

### PR DESCRIPTION
**Pull request by DrakeLin**
## 🥞 Stacked PR
Use this [link](https://github.com/delta-io/delta-kernel-rs/pull/2521/files) to review incremental changes.
- [stack/test-table-builder-last-checkpoint-hint](https://github.com/delta-io/delta-kernel-rs/pull/2521) [[Files changed](https://github.com/delta-io/delta-kernel-rs/pull/2521/files)]

---------
## What changes are proposed in this pull request?

Adds a `last_checkpoint_hint` axis to `LogState` with three states (`Present` / `Missing` / `Stale`). `Stale` rewrites the hint to point at the lowest checkpoint, forcing the reader's listing-fallback recovery to discover the actual latest checkpoint.

## How was this change tested?

unit tests

### CODE REVIEW

File: test-utils/src/table_builder.rs [hunk 1]
```diff
@@ -208,6 +269,15 @@ impl LogState {
             Self::with_latest_version(N).with_checkpoint_at([N - 5]),
             // multi-checkpoint: newer supersedes older
             Self::with_latest_version(N).with_checkpoint_at([N - 5, N]),
+            // mid-stream checkpoint with hint missing: listing fallback discovers it
+            Self::with_latest_version(N)
+                .with_checkpoint_at([N - 5])
+                .keep_last_checkpoint_hint(false),
```
Comment (scottsand-db) [inline test-utils/src/table_builder.rs]: nit: is this really any clearer or better than `.with_last_checkpoint_hint(LastCheckpointHintState::Missing)` ?

since below we use LastCheckpointHintState

File: test-utils/src/table_builder.rs [hunk 2]
```diff
@@ -228,6 +298,23 @@ impl LogState {
...
     }
+
+    /// State of the `_last_checkpoint` hint file on the built table.
+    pub(crate) fn last_checkpoint_hint(&self) -> LastCheckpointHintState {
+        self.last_checkpoint_hint
+    }
+}
+
+/// Validate up-front in `build()` so panic messages surface synchronously rather than
+/// through the worker-thread join boundary in [`block_on_sync`].
```
Comment (scottsand-db) [inline test-utils/src/table_builder.rs]: Comment isn't super helpful? I don't need to know the nitty gritty worker-thread boundary conditions in order to understand this fn

I feel like we don't even need a comment here tbh

File: test-utils/src/table_builder.rs [hunk 3]
```diff
@@ -940,6 +1036,42 @@ impl TestTableBuilder {
...
+                    Ok(()) | Err(ObjectStoreError::NotFound { .. }) => {}
+                    Err(e) => return Err(delta_kernel::Error::from(e)),
+                }
+            }
+            LastCheckpointHintState::Stale => {
+                // Rewrite the hint to point at the LOWEST checkpoint (rather than the
+                // highest one kernel just wrote). The reader will follow the hint, list
+                // forward from that version, find the actual highest checkpoint, and
+                // log an info about the version mismatch -- exactly the "stale older
+                // hint" recovery path the spec calls out.
```
Comment (scottsand-db) [inline test-utils/src/table_builder.rs]: IMO this isn't the place to describe what the reader will do when it sees the stale hint. That should be on `LastCheckpointHintState::Stale`.

Here, it is succinct and helpful to just say we rewrite using the lowest checkpoint

File: test-utils/src/table_builder.rs [hunk 4]
```diff
@@ -940,6 +1036,42 @@ impl TestTableBuilder {
...
+                let stale_version = self
+                    .log_state
+                    .checkpoints_at()
+                    .first()
+                    .copied()
+                    .expect("validate_log_state should have caught Stale + 0 checkpoints");
+                let hint_path = Path::from("_delta_log/_last_checkpoint");
+                let resolved = crate::resolve_table_path(table_root, &hint_path)?;
+                let body = format!(r#"{{"version":{stale_version},"size":1}}"#);
```
Comment (scottsand-db) [inline test-utils/src/table_builder.rs]: hmm ... doesn't the file contain a bunch more info?
Can't we .... _read_ the file as we are writing? Save that? And then just re-write those contents later? that way we 100% match what kernel writes?
- write checkpoint 5
- copy `_last_checkpoint` content
- stuff
- write checkpoint 10. this overwrites it
- we over-write it outselves with that saved content
- we can use the read bytes storage handler API?

**Review by scottsand-db — CHANGES_REQUESTED** [review]
Looks great! Some comments

File: test-utils/src/table_builder.rs (hunk 4 above)
Comment (DrakeLin) [reply test-utils/src/table_builder.rs]: good point, will do that

File: test-utils/src/table_builder.rs [hunk 5]
```diff
@@ -1664,11 +1823,13 @@ mod tests {
         assert_eq!(log_state.latest_version(), expected);
     }
 
-    /// Asserts that every checkpoint version in `log_state.checkpoints_at()`
-    /// produced an on-disk checkpoint file. Snapshot rebuilds can silently
-    /// succeed via JSON-replay fallback even when a checkpoint is missing, so
-    /// any LogState-driven test should call this to verify the fixture's
-    /// contract before asserting anything else.
+    /// Asserts that the on-disk log directory matches the contract advertised by
```
Comment (scottsand-db) [inline test-utils/src/table_builder.rs]: IMO this fn doc can be simplified and made more high level. we can just have deeper per-validation comments inside the fn

File: test-utils/src/table_builder.rs [hunk 6]
```diff
@@ -914,12 +1000,23 @@ impl TestTableBuilder {
                 self.clustering_columns.iter().map(|s| s.as_str()),
             ));
         }
+        let hint_state = self.log_state.last_checkpoint_hint();
+        let hint_path = Path::from("_delta_log/_last_checkpoint");
+        let resolved_hint_path = crate::resolve_table_path(table_root, &hint_path)?;
+        // For Stale, snapshot the hint kernel writes after the lowest checkpoint and
```
Comment (scottsand-db) [inline test-utils/src/table_builder.rs]: `snapshot` might be a confusing term here

File: test-utils/src/table_builder.rs [hunk 7]
```diff
@@ -914,12 +1000,23 @@ impl TestTableBuilder {
...
+        // restore it at the end. This way the hint bytes exactly match what kernel
```
Comment (scottsand-db) [inline test-utils/src/table_builder.rs]: `This way the hint bytes exactly match what kernel would have written, with no fabricated JSON.` -- again sorry but seems like AI comment. I gave feedback for X, this comment says "we don't do X so that ...".

IMO it's clear enough without this part of the comment

File: test-utils/src/table_builder.rs [hunk 8]
```diff
@@ -857,7 +937,13 @@ impl TestTableBuilder {
     /// runtime on a background thread to avoid panicking on nested runtimes.
     /// Propagates I/O, create-table, commit, checkpoint, and CRC errors from the
     /// underlying write path.
+    ///
+    /// # Panics
+    /// Panics if [`LastCheckpointHintState::Stale`] is paired with fewer than two
+    /// checkpoints. Validated up-front so the message surfaces directly rather than
```
Comment (scottsand-db) [inline test-utils/src/table_builder.rs]: `Validated up-front so the message surfaces directly rather than` can delete this last sentence

File: test-utils/src/table_builder.rs [hunk 9]
```diff
@@ -224,6 +270,15 @@ impl LogState {
             Self::with_latest_version(N)
                 .with_checkpoint_at([N - 5])
                 .with_cleanup_commits_before(N - 5),
+            // mid-stream checkpoint with hint missing: listing fallback discovers it
+            Self::with_latest_version(N)
+                .with_checkpoint_at([N - 5])
+                .with_last_checkpoint_hint(LastCheckpointHintState::Missing),
```
Comment (scottsand-db) [inline test-utils/src/table_builder.rs:276]: should we also test with_last_checkpoint_hint missing/stale and with cleanup commits?

in general, have some commits cleaned up is the most common state of any delta table -- we should definintely test this

**Review by scottsand-db — APPROVED** [review]
LGTM but please test `with_cleanup_commits_before` + Stale

**Review by Jameson-Crate — APPROVED** [review]
looks good to me
